pytorch_lightning/trainer/training_loop.py

```diff
@@ -515,8 +516,22 @@ def run_training_batch(self, batch, batch_idx):
         for split_idx, split_batch in enumerate(splits):
             self.split_idx = split_idx
 
+            def get_optimizers_iterable():
+                if not self.optimizer_frequencies:
+                    return enumerate(self.optimizers)
+
+                optimizer_freq_cumsum = np.cumsum(self.optimizer_frequencies)
```

Comment:
not sure if I follow here, you want to filter (select) just optimizes if the actual batch index is divided by its frequency, right?
 so having optA with freq=2 and optB with freq=3 you would apply:
- it=0 >> optA, optB
- it=1 >> -
- it=2 >> optA
- it=3 >> optB
- ...

Reply:
I'll explain:

```py
optimizer_frequencies = [2, 3]
optimizer_freq_cumsum = [2, 5]
optimizers_loop_length = 5
```

The loop finds the first object in `optimizer_freq_cumsum` that is bigger than `current_place_in_loop`:

```py
self.total_batch_idx = 0
current_place_in_loop = 0 % 5   # == 0
# v = 2, v > 0 == True
opt_idx = 0
...
self.total_batch_idx = 1
current_place_in_loop = 1
# v = 2, v > 1 == True
opt_idx = 0
...
self.total_batch_idx = 2
current_place_in_loop = 2
# v = 2, v > 2 == False
# v = 5, v > 2 == True
opt_idx = 1
...
self.total_batch_idx = 3
current_place_in_loop = 3
# v = 2, v > 3 == False
# v = 5, v > 3 == True
opt_idx = 1
...
self.total_batch_idx = 4
current_place_in_loop = 4
# v = 2, v > 4 == False
# v = 5, v > 4 == True
opt_idx = 1
...
self.total_batch_idx = 5
current_place_in_loop = 5 % 5  # == 0
# v = 2, v > 0 == True
opt_idx = 0
...
self.total_batch_idx = 6
current_place_in_loop = 1
# v = 2, v > 1 == True
opt_idx = 0
...

Reply:
So frequency amounts to the number of sequential operations of each optimizer.
You can see that optimizer 0 is chosen 2 times in a row,
And then optimizer 1 is chosen 3 times in a row.